criu/pidfd.c

```diff
@@ -0,0 +1,229 @@
+		return -1;
+
+	/*
+	* Check if the pid pidfd refers to is part of process tree
+	* This ensures the process will exist on restore.
+	*/
+	if (pidfd_info.pid != -1 && !pstree_item_by_real(pidfd_info.pid)) {
+		pr_err("pidfd pid %d is not a part of process tree..",
+			pidfd_info.pid);
+		return -1;
+	}
+
+	pidfd_info.pidfe.id = id;
+	pidfd_info.pidfe.flags = (p->flags & ~O_RDWR);
+	pidfd_info.pidfe.fown = (FownEntry *)&p->fown;
+
+	fe.type = FD_TYPES__PIDFD;
+	fe.id = pidfd_info.pidfe.id;
+	fe.pidfd = &pidfd_info.pidfe;
+
+	pr_info_pidfd("Dumping", &pidfd_info.pidfe);
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+const struct fdtype_ops pidfd_dump_ops = {
+	.type = FD_TYPES__PIDFD,
+	.dump = dump_one_pidfd,
+};
+
+static int pidfd_open(pid_t pid, int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int create_tmp_process(void)
+{
+	int tmp_process;
+	tmp_process = fork();
+	if (tmp_process < 0) {
+		pr_perror("Could not fork");
+		return -1;
+	} else if (tmp_process == 0) {
+		while(1)
+			sleep(1);
+	}
+	return tmp_process;
+}
+
+static int open_one_pidfd(struct file_desc *d, int *new_fd)
+{
+	struct pidfd_info *info;
+	struct dead_pidfd *dead;
+	int pidfd;
+
+	info = container_of(d, struct pidfd_info, d);
+
+	if (info->pidfe->nspid != -1) {
+		pidfd = pidfd_open(info->pidfe->nspid, info->pidfe->flags);
+		if (pidfd < 0)
+			goto err_close;
+		goto out;
+	}
+
+	dead = lookup_dead_pidfd(info->pidfe->ino);
+	BUG_ON(!dead);
+
+	dead->count--;
+	if (dead->pid == -1) {
+		dead->pid = create_tmp_process();
+		if (dead->pid < 0)
+			goto err_close;
+	}
+
+	pidfd = pidfd_open(dead->pid, info->pidfe->flags);
+	if (pidfd < 0)
+		goto err_close;
+
+	if (dead->count == 0) {
+		int status;
+
+		kill(dead->pid, SIGKILL);
+		waitpid(dead->pid, &status, 0);
+		if (WTERMSIG(status) != SIGKILL)
+			goto err_close;
+		hlist_del(&dead->hash);
+		xfree(dead);
+	}
+out:
+	if (rst_file_params(pidfd, info->pidfe->fown, info->pidfe->flags)) {
+		goto err_close;
+	}
+
+	*new_fd = pidfd;
+	return 0;
+
+err_close:
+	free_dead_pidfd_hash();
+	pr_perror("Can't create pidfd %#08x NSpid: %d flags: %u",
+	   info->pidfe->id, info->pidfe->nspid, info->pidfe->flags);
+	return -1;
+}
+
+static struct file_desc_ops pidfd_desc_ops = {
+	.type = FD_TYPES__PIDFD,
+	.open = open_one_pidfd
+};
+
+static int collect_one_pidfd(void *obj, ProtobufCMessage *msg, struct cr_img *i)
+{
+	struct dead_pidfd *dead;
+	struct pidfd_info *info = obj;
+
+	info->pidfe = pb_msg(msg, PidfdEntry);
+	pr_info_pidfd("Collected ", info->pidfe);
+
+	if (info->pidfe->nspid != -1)
+		goto out;
+
+	dead = lookup_dead_pidfd(info->pidfe->ino);
+	if (dead) {
+		dead->count++;
+		goto out;
+	}
+
+	dead = xmalloc(sizeof(*dead));
+	if (!dead)
+		return -1;
+
+	INIT_HLIST_NODE(&dead->hash);
+	dead->ino = info->pidfe->ino;
+	dead->count = 1;
+	dead->pid = -1;
+
+	hlist_add_head(&dead->hash, &dead_pidfd_hash[dead->ino % DEAD_PIDFD_HASH_SIZE]);
```

Comment:
why do we need more than one dead process?

Reply:
We need a unique process for each unique inode, so that pidfd comparisons don't break after restore.
All pidfds that point to the same process have the same inode number.
See also https://github.com/torvalds/linux/commit/cb12fd8e0dabb9a1c8aef55a6a41e2c255fcdf4b

Reply:
make sense.